In [92]:
print("all ok")

all ok


In [93]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

In [94]:
documents=[
    "People watch movie",
    "people watch cricket",
    "people like movie",
    "people like cricket"
]

In [95]:
tokens=[sentence.lower().split() for sentence in documents]

In [96]:
tokens

[['people', 'watch', 'movie'],
 ['people', 'watch', 'cricket'],
 ['people', 'like', 'movie'],
 ['people', 'like', 'cricket']]

In [97]:
all_words=[[word] for sentence in tokens for word in sentence]

👉 This is called nested list comprehension

🔴 Let’s break it slowly:
First:
for sentence in tokens

👉 takes one sentence at a time:

['people', 'watch', 'movie']
Then:
for word in sentence

👉 takes each word:

'people', 'watch', 'movie'
🔹 Final:
[word]

👉 puts each word inside a list

✅ Final Output:
[
 ['people'],
 ['watch'],
 ['movie'],
 ['people'],
 ['watch'],
 ['cricket'],
 ['people'],
 ['like'],
 ['movie'],
 ['people'],
 ['like'],
 ['cricket']
]
🧠 Simple Meaning:

👉 You are converting:

list of sentences → list of words → list of single-word lists
💡 Why do this?

This format is used in:

Bag of Words (BOW)
Word2Vec
NLP models

Because many models expect:
👉 each word as a separate input

🔹 Easier way to understand:

Same thing using normal loop:

all_words = []

for sentence in tokens:
    for word in sentence:
        all_words.append([word])

In [98]:
all_words=[[word] for sentence in tokens for word in sentence]

In [99]:
all_words

[['people'],
 ['watch'],
 ['movie'],
 ['people'],
 ['watch'],
 ['cricket'],
 ['people'],
 ['like'],
 ['movie'],
 ['people'],
 ['like'],
 ['cricket']]

sparce_output=False means the encoder returns a numpy array instead of a sparce matrix .

In [100]:
encoder = OneHotEncoder(sparse_output=False)

In [101]:
encoder.fit(all_words)

OneHotEncoder(sparse_output=False)

In [102]:
print("Vocabulary:",encoder.categories_[0])

Vocabulary: ['cricket' 'like' 'movie' 'people' 'watch']


In [103]:
encoder.categories_

[array(['cricket', 'like', 'movie', 'people', 'watch'], dtype=object)]

In [104]:
encoder.categories_[0]

array(['cricket', 'like', 'movie', 'people', 'watch'], dtype=object)

In [105]:
tokens

[['people', 'watch', 'movie'],
 ['people', 'watch', 'cricket'],
 ['people', 'like', 'movie'],
 ['people', 'like', 'cricket']]

In [106]:
for sentence in tokens:
    encoded_sentence=encoder.transform([[word] for word in sentence])
    print("\nSentence:",sentence)
    print(encoded_sentence)


Sentence: ['people', 'watch', 'movie']
[[0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 1. 0. 0.]]

Sentence: ['people', 'watch', 'cricket']
[[0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0.]]

Sentence: ['people', 'like', 'movie']
[[0. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]

Sentence: ['people', 'like', 'cricket']
[[0. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0.]
 [1. 0. 0. 0. 0.]]


BOW

In [107]:
from sklearn.feature_extraction.text import CountVectorizer

In [108]:
documents = ["people watch movie and watch movie again", 
             "people watch cricket and watch cricket",
             "people like movie and like movie a lot",
             "people like cricket"]

In [109]:
bow = CountVectorizer()

In [110]:
bow_matrix = bow.fit_transform(documents)

In [111]:
bow_matrix

<4x8 sparse matrix of type '<class 'numpy.int64'>'
	with 17 stored elements in Compressed Sparse Row format>

In [112]:
print("Bow matrix:\n",bow_matrix.toarray())

Bow matrix:
 [[1 1 0 0 0 2 1 2]
 [0 1 2 0 0 0 1 2]
 [0 1 0 2 1 2 1 0]
 [0 0 1 1 0 0 1 0]]


In [113]:
print("Vocabulary:",bow.get_feature_names_out())

Vocabulary: ['again' 'and' 'cricket' 'like' 'lot' 'movie' 'people' 'watch']


In [114]:
bow.get_feature_names_out()

array(['again', 'and', 'cricket', 'like', 'lot', 'movie', 'people',
       'watch'], dtype=object)

In [115]:
# bow_matrix=bow.fit_transform(document)


In [116]:
document = ["lion is the king of jungle"]
bow.transform(document).toarray()

array([[0, 0, 0, 0, 0, 0, 0, 0]], dtype=int64)

In [117]:
bow_matrix.toarray()

array([[1, 1, 0, 0, 0, 2, 1, 2],
       [0, 1, 2, 0, 0, 0, 1, 2],
       [0, 1, 0, 2, 1, 2, 1, 0],
       [0, 0, 1, 1, 0, 0, 1, 0]], dtype=int64)

In [118]:
documents

['people watch movie and watch movie again',
 'people watch cricket and watch cricket',
 'people like movie and like movie a lot',
 'people like cricket']

In [119]:
bow.transform(documents).toarray()

array([[1, 1, 0, 0, 0, 2, 1, 2],
       [0, 1, 2, 0, 0, 0, 1, 2],
       [0, 1, 0, 2, 1, 2, 1, 0],
       [0, 0, 1, 1, 0, 0, 1, 0]], dtype=int64)

In [120]:
document=["Lion is the king of jungle"]

In [121]:
bow.transform(document)

<1x8 sparse matrix of type '<class 'numpy.int64'>'
	with 0 stored elements in Compressed Sparse Row format>

Aim:Convert:Data:Number

Basic Example
data = [['red'], ['blue'], ['green']]

encoder = OneHotEncoder(sparse_output=False)
encoded = encoder.fit_transform(data)

print(encoded)


✅ Output:
[
 [0,0,1],   # red
 [1,0,0],   # blue
 [0,1,0]    # green
]
🔹 Important Methods
1. fit()
encoder.fit(data)

👉 Learns all unique values (vocabulary)

2. transform()
encoder.transform([['red']])

👉 Converts to vector

3. fit_transform()
encoder.fit_transform(data)

👉 Shortcut (fit + transform)

4. categories_
encoder.categories_

👉 Shows all unique values:

[array(['blue', 'green', 'red'], dtype=object)]
🔹 Why 2D input?
[['red'], ['blue']] ✅
['red', 'blue'] ❌

👉 Because sklearn expects:
👉 rows = samples, columns = features

🔹 In your NLP case
all_words = [['people'], ['watch'], ['movie'], ...]

👉 Each word = one sample
👉 Encoder learns vocabulary
👉 Converts each word → vector

TF-IDF Practical

In [122]:
from sklearn.feature_extraction.text import TfidfVectorizer


In [123]:
documents = ["people watch movie and watch movie again", 
             "people watch cricket and watch cricket",
             "people like movie and like movie a lot",
             "people like cricket"]

In [124]:
tf_idf=TfidfVectorizer()

In [125]:
tf_idf_vector = tf_idf.fit_transform(documents)

In [126]:
tf_idf_vector

<4x8 sparse matrix of type '<class 'numpy.float64'>'
	with 17 stored elements in Compressed Sparse Row format>

In [127]:
tf_idf_vector.toarray()

array([[0.38771136, 0.24747114, 0.        , 0.        , 0.        ,
        0.61135218, 0.20232387, 0.61135218],
       [0.        , 0.2684707 , 0.66322944, 0.        , 0.        ,
        0.        , 0.21949239, 0.66322944],
       [0.        , 0.24747114, 0.        , 0.61135218, 0.38771136,
        0.61135218, 0.20232387, 0.        ],
       [0.        , 0.        , 0.64043405, 0.64043405, 0.        ,
        0.        , 0.42389674, 0.        ]])

In [128]:
for vector in tf_idf_vector.toarray():
    print(vector)

[0.38771136 0.24747114 0.         0.         0.         0.61135218
 0.20232387 0.61135218]
[0.         0.2684707  0.66322944 0.         0.         0.
 0.21949239 0.66322944]
[0.         0.24747114 0.         0.61135218 0.38771136 0.61135218
 0.20232387 0.        ]
[0.         0.         0.64043405 0.64043405 0.         0.
 0.42389674 0.        ]


In [129]:
tf_idf.get_feature_names_out()

array(['again', 'and', 'cricket', 'like', 'lot', 'movie', 'people',
       'watch'], dtype=object)

In [130]:
print("Vocabulary:", tf_idf.get_feature_names_out())

Vocabulary: ['again' 'and' 'cricket' 'like' 'lot' 'movie' 'people' 'watch']


Assignment1

✅ STEP 1: Collect Dataset (Reviews)
Option A (Easy for beginner ⭐)

👉 Don’t scrape (to save time), use sample data first

In [145]:
import requests
from bs4 import BeautifulSoup

url = "https://www.amazon.in/product-reviews/B0FGY54N5T"

headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36",
    "Accept-Language": "en-US,en;q=0.9"
}

response = requests.get(url, headers=headers)

soup = BeautifulSoup(response.content, "html.parser")

print(soup.title)

<title dir="ltr">Amazon Sign-In</title>


In [146]:
from bs4 import BeautifulSoup

with open("page.html", "r", encoding="utf-8") as f:
    soup = BeautifulSoup(f, "html.parser")

print(soup.title)

<title>Amazon.in:Customer reviews: Youthnic Women&amp;#39;s Poly Rayon Printed Alia Cut Peplum Top | Top for Office | Women&amp;#39;s Tunics | Short Kurti Women| Regular Fit</title>


In [156]:
from bs4 import BeautifulSoup

with open("page.html", "r", encoding="utf-8") as f:
    soup = BeautifulSoup(f, "html.parser")

reviews = soup.find_all("span", {"data-hook": "review-body"})

review_list = []

for r in reviews:
    # print(r)
    # print("---------------------------------------")
    review_list.append(r.text.strip())

print(review_list)
print(len(review_list))

['The quality is amazingg !!! The fabric is pure cotton, so very well for summer and the fitting is greatttt……. Great for everyday wear.', "It's very good.. Worth the price..", 'Beautiful ethnic top, great qualityI really loved this women’s cotton blend bandhani print peplum top❤️', 'Nice dress', 'Ok,', 'The product was received by me without any tag ... The same when returned due to size difficulty the Person said there is no tag ... But none was received ... This is very disappointing never happened with me ... Least expected', 'Beautiful - love the shade! It looks great on jeans! Palazzo or dhoti!', 'Not so good', "It's good", 'Great purchase. I like both the material and fittings.Go for it girls']
10


In [ ]:
# STEP 1: Lowercase
review_list = [r.lower() for r in review_list]

print(review_list)

['the quality is amazingg !!! the fabric is pure cotton, so very well for summer and the fitting is greatttt……. great for everyday wear.', "it's very good.. worth the price..", 'beautiful ethnic top, great qualityi really loved this women’s cotton blend bandhani print peplum top❤️', 'nice dress', 'ok,', 'the product was received by me without any tag ... the same when returned due to size difficulty the person said there is no tag ... but none was received ... this is very disappointing never happened with me ... least expected', 'beautiful - love the shade! it looks great on jeans! palazzo or dhoti!', 'not so good', "it's good", 'great purchase. i like both the material and fittings.go for it girls']


In [ ]:
# STEP 2: Remove punctuation
import re

def remove_punctuation(text):
    return re.sub(r'[^\w\s]', '', text)

review_list = [remove_punctuation(r) for r in review_list]

print(review_list)

['the quality is amazingg  the fabric is pure cotton so very well for summer and the fitting is greatttt great for everyday wear', 'its very good worth the price', 'beautiful ethnic top great qualityi really loved this womens cotton blend bandhani print peplum top', 'nice dress', 'ok', 'the product was received by me without any tag  the same when returned due to size difficulty the person said there is no tag  but none was received  this is very disappointing never happened with me  least expected', 'beautiful  love the shade it looks great on jeans palazzo or dhoti', 'not so good', 'its good', 'great purchase i like both the material and fittingsgo for it girls']


In [159]:
# STEP 3: Tokenization
tokens_list = [r.split() for r in review_list]

print(tokens_list)

[['the', 'quality', 'is', 'amazingg', 'the', 'fabric', 'is', 'pure', 'cotton', 'so', 'very', 'well', 'for', 'summer', 'and', 'the', 'fitting', 'is', 'greatttt', 'great', 'for', 'everyday', 'wear'], ['its', 'very', 'good', 'worth', 'the', 'price'], ['beautiful', 'ethnic', 'top', 'great', 'qualityi', 'really', 'loved', 'this', 'womens', 'cotton', 'blend', 'bandhani', 'print', 'peplum', 'top'], ['nice', 'dress'], ['ok'], ['the', 'product', 'was', 'received', 'by', 'me', 'without', 'any', 'tag', 'the', 'same', 'when', 'returned', 'due', 'to', 'size', 'difficulty', 'the', 'person', 'said', 'there', 'is', 'no', 'tag', 'but', 'none', 'was', 'received', 'this', 'is', 'very', 'disappointing', 'never', 'happened', 'with', 'me', 'least', 'expected'], ['beautiful', 'love', 'the', 'shade', 'it', 'looks', 'great', 'on', 'jeans', 'palazzo', 'or', 'dhoti'], ['not', 'so', 'good'], ['its', 'good'], ['great', 'purchase', 'i', 'like', 'both', 'the', 'material', 'and', 'fittingsgo', 'for', 'it', 'girls']]


In [160]:
# STEP 4: Remove Stopwords (Optional but good)
from nltk.corpus import stopwords
import nltk

nltk.download('stopwords')

stop_words = set(stopwords.words('english'))

tokens_list = [
    [word for word in tokens if word not in stop_words]
    for tokens in tokens_list
]

print(tokens_list)

[['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear'], ['good', 'worth', 'price'], ['beautiful', 'ethnic', 'top', 'great', 'qualityi', 'really', 'loved', 'womens', 'cotton', 'blend', 'bandhani', 'print', 'peplum', 'top'], ['nice', 'dress'], ['ok'], ['product', 'received', 'without', 'tag', 'returned', 'due', 'size', 'difficulty', 'person', 'said', 'tag', 'none', 'received', 'disappointing', 'never', 'happened', 'least', 'expected'], ['beautiful', 'love', 'shade', 'looks', 'great', 'jeans', 'palazzo', 'dhoti'], ['good'], ['good'], ['great', 'purchase', 'like', 'material', 'fittingsgo', 'girls']]


[nltk_data] Downloading package stopwords to C:\Users\Shraddha
[nltk_data]     Chaudhari\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


TASK 2: Vocabulary Creation (VERY EASY)
🧠 What is Vocabulary?

👉 Vocabulary = all unique words from your data

🔍 Example

If your cleaned data is:

[
 ['quality', 'amazing'],
 ['good'],
 ['nice', 'dress']
]

👉 Vocabulary becomes:

['quality', 'amazing', 'good', 'nice', 'dress']

👉 No duplicates ❗

In [166]:
from collections import Counter

all_words = []

for tokens in tokens_list:
    all_words.extend(tokens)
    # print(tokens)
    print(all_words)
    
print("----------------------------------------")
vocab = Counter(all_words)

print("Vocabulary:", vocab)
print("Vocabulary size:", len(vocab))
print("Top 5 words:", vocab.most_common(5))

['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear']
['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear', 'good', 'worth', 'price']
['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear', 'good', 'worth', 'price', 'beautiful', 'ethnic', 'top', 'great', 'qualityi', 'really', 'loved', 'womens', 'cotton', 'blend', 'bandhani', 'print', 'peplum', 'top']
['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear', 'good', 'worth', 'price', 'beautiful', 'ethnic', 'top', 'great', 'qualityi', 'really', 'loved', 'womens', 'cotton', 'blend', 'bandhani', 'print', 'peplum', 'top', 'nice', 'dress']
['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear', 'good', 'worth', 'pri

In [ ]:
# TASK 3: Feature Engineering
# Now we convert words → numbers

# 🔹 1. One Hot Encoding (Simple)

# 👉 Idea:

# If word present → 1
# If not → 0

vocab_list = list(vocab.keys())
print("Vocabulary List:(Unique words)", vocab_list)

def one_hot(doc):
    return [1 if word in doc else 0 for word in vocab_list]

ohe_vectors = [one_hot(doc) for doc in tokens_list]

print(ohe_vectors)

Vocabulary List: ['quality', 'amazingg', 'fabric', 'pure', 'cotton', 'well', 'summer', 'fitting', 'greatttt', 'great', 'everyday', 'wear', 'good', 'worth', 'price', 'beautiful', 'ethnic', 'top', 'qualityi', 'really', 'loved', 'womens', 'blend', 'bandhani', 'print', 'peplum', 'nice', 'dress', 'ok', 'product', 'received', 'without', 'tag', 'returned', 'due', 'size', 'difficulty', 'person', 'said', 'none', 'disappointing', 'never', 'happened', 'least', 'expected', 'love', 'shade', 'looks', 'jeans', 'palazzo', 'dhoti', 'purchase', 'like', 'material', 'fittingsgo', 'girls']
[[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0

In [ ]:
# let say example 
# tokens_list = [
#  ['quality', 'amazing'],
#  ['good'],
#  ['nice', 'dress']
# ]

# doc = ['quality', 'amazing']

# vocab_list = all unique words from ALL reviews

# From tokens_list:

# ['quality', 'amazing']
# ['good']
# ['nice', 'dress']

# 👉 Combine:
# ✅ Final vocab_list
# vocab_list = ['quality', 'amazing', 'good', 'nice', 'dress']

# 👉 Order may change, but words will be same

# ⚠️ Important Point

# 👉 doc is only one review
# 👉 But vocab_list is from all reviews

# 🔁 Step 2: Apply One Hot Encoding

# Now:

# doc = ['quality', 'amazing']
# vocab_list = ['quality', 'amazing', 'good', 'nice', 'dress']

# 🔍 Check each word
# word (from vocab)	in doc?	output
# # quality	✅ yes	1
# amazing	✅ yes	1
# good	❌ no	0
# nice	❌ no	0
# dress	❌ no	0
# ✅ Final Output
# [1, 1, 0, 0, 0]

SyntaxError: invalid syntax (1441288707.py, line 1)

In [ ]:
# Your code (list comprehension)
def one_hot(doc):
    return [1 if word in doc else 0 for word in vocab_list]

ohe_vectors = [one_hot(doc) for doc in tokens_list]
print(ohe_vectors)


# 🔁 Same code using SIMPLE loops (easy to understand)
ohe_vectors = []

for doc in tokens_list:          # loop over each review
    vector = []                  # store result for one review
    
    for word in vocab_list:      # check each word in vocab
        
        if word in doc:
            vector.append(1)
        else:
            vector.append(0)
    
    ohe_vectors.append(vector)   # add final vector

print(ohe_vectors)




# 🔍 Example Flow

# For:

# doc = ['good', 'product']
# vocab_list = ['good', 'product', 'bad']

# 👉 vector becomes:

# [1, 1, 0]

<!-- Chatgpt -->


**Please give a small example with fewer words so that I can understand this logic.**

---

### 🧠 Very Small Example (2 reviews only)

#### ✅ Step 1: Data

```python
tokens_list = [
    ['good', 'product'],
    ['bad', 'product']
]
```

---

#### ✅ Step 2: Vocabulary

👉 Take all unique words:

```python
vocab_list = ['good', 'product', 'bad']
```

---

#### ✅ Step 3: One Hot Encoding Code

```python
def one_hot(doc):
    return [1 if word in doc else 0 for word in vocab_list]

ohe_vectors = [one_hot(doc) for doc in tokens_list]

print(ohe_vectors)
```

---

### 🔍 Step-by-step Understanding

#### 🔹 First Document

```python
doc = ['good', 'product']
```

| Word    | In Doc? | Output |
| ------- | ------- | ------ |
| good    | Yes     | 1      |
| product | Yes     | 1      |
| bad     | No      | 0      |

👉 Output:

```
[1, 1, 0]
```

---

#### 🔹 Second Document

```python
doc = ['bad', 'product']
```

| Word    | In Doc? | Output |
| ------- | ------- | ------ |
| good    | No      | 0      |
| product | Yes     | 1      |
| bad     | Yes     | 1      |

👉 Output:

```
[0, 1, 1]
```

---

### ✅ Final Output

```
[
 [1, 1, 0],
 [0, 1, 1]
]
```

---

### 🧠 Simple Meaning

* Each row = one review
* Each column = one word

---

### 💡 Easy Trick to Remember

👉 Ask for each word:
**“Is this word in this review?”**

* Yes → 1
* No → 0

---

### 🎯 Your Turn (Very Easy)

```python
tokens_list = [
    ['good'],
    ['bad']
]

vocab_list = ['good', 'bad']
```

👉 What will be the output?


In [169]:
# 🧠 What is Bag of Words?

# 👉 It counts how many times each word appears

# 🔍 Example

# Sentence:

# good good product

# 👉 Output:

# good → 2  
# product → 1
# 🧪 Code (Use your review_list)

from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()

bow = cv.fit_transform(review_list)

print(bow.toarray())
print(cv.get_feature_names_out())

# 🧠 What is happening?
# Line	Meaning
# CountVectorizer()	create BoW model
# fit_transform()	learn + convert
# toarray()	show numbers
# 📊 Output Example
# [[0 1 0 2]
#  [1 0 1 0]]

# 👉 Words:

# ['bad', 'good', 'nice', 'product']
# 🎯 DONE ✅ (BoW complete)

[[1 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 2 0 0 1 1 0 3 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 1 1 0 3 0 0 0 0 1
  0 1 1 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0
  0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1
  0 0 0 0 0 0 0 1]
 [0 0 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1
  0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 2 0
  0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 0]
 [0 0 1 0 0 0 0 1 1 0 0 1 1 0 1 0 0 1 0 0 0 0 0 0 0 0 1 2 0 0 0 1 0 0 0 0
  0 2 1 0 1 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 2 1 1 1 0 1 0 0 2 3 1 1 1 0 1
  2 0 0 1 1 1 0 0

In [ ]:
# 🌱 Very Small Dataset (Super Easy)

# Let’s take 3 simple sentences:

# re_list = [
#     "good product",
#     "good quality",
#     "bad product"
# ]
# 🧠 Step 1: Vocabulary

# 👉 Unique words:

# ['bad', 'good', 'product', 'quality']
# 🧪 Step 2: Apply Bag of Words


re_list = [
    "good product",
    "good quality",
    "bad product"
]

from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()

bow = cv.fit_transform(re_list)

print(cv.get_feature_names_out())
print(bow.toarray())

# 📊 Output
# Words (Vocabulary order):
# ['bad', 'good', 'product', 'quality']
# Matrix:
# [
#  [0 1 1 0]   → "good product"
#  [0 1 0 1]   → "good quality"
#  [1 0 1 0]   → "bad product"
# ]
# 🧠 Understand Row by Row
# 🔹 Sentence 1: "good product"
# [0 1 1 0]
# word	count
# bad	0
# good	1
# product	1
# quality	0
# 🔹 Sentence 2: "good quality"
# [0 1 0 1]
# 🔹 Sentence 3: "bad product"
# [1 0 1 0]
# 🎯 Key Idea (VERY IMPORTANT)

# 👉 Each row = one sentence
# 👉 Each column = one word
# 👉 Value = how many times word appears

# 💡 Super Simple Formula

# 👉 BoW = Count of words

# 🧪 One More Tiny Example
# review_list = [
#     "good good product",
# ]

# 👉 Output:

# ['good', 'product']

# [2 1]

# 👉 Because:

# good → 2 times
# product → 1 time
# 🚀 Your Turn

# Try this small dataset yourself and tell me:

# 👉 Do you understand how numbers are coming?

# Next step (very similar):
# 👉 TF-IDF (same idea but smarter) 👍

['bad' 'good' 'product' 'quality']
[[0 1 1 0]
 [0 1 0 1]
 [1 0 1 0]]



🧠 What does that mean?

👉 In Bag of Words:

Each word = one feature (column)
We ignore:
❌ grammar
❌ order
❌ meaning
🔍 Example again
review_list = [
    "good good product"
]

👉 Vocabulary:

['good', 'product']

👉 Output:

[2 1]
🧠 Why it is called word-level?

Because:

👉 We are focusing on individual words only

word	count
good	2
product	1

📊 So:
Type	Explanation
Word-level	counts words individually
Not sentence-level	ignores structure
Not semantic	ignores meaning
💡 Simple Line to Remember

👉 Bag of Words = word-level frequency representation

🚀 Small Extra (for understanding)

👉 These are ALSO word-level:

One Hot Encoding ✅
Bag of Words ✅
TF-IDF ✅

👉 These are NOT:

Word2Vec ❌ (captures meaning)
BERT ❌ (understands context)
🎯 Final Answer for Assignment

👉 You can write:

“Bag of Words is a word-level representation where each word is treated as an independent feature, and the value represents the frequency of that word in the document.”

👉 It may look like One Hot Encoding
But difference is:

Feature	  OHE	          BoW
Value	  0 or 1      only	0,1,2,3…
Meaning	  presence	   frequency

🎯 Final Answer for Assignment

You can write:

“Bag of Words is a word-level representation that captures the frequency of each word in a document. In this example, frequencies are mostly 0 or 1 because words appear only once.”

TF-IDF

In [172]:
# 🧠 What is TF-IDF?

# 👉 It gives importance to words

# common words → low value
# rare words → high value
# 🔍 Example

# If "good" appears in all reviews → ❌ less important
# If "disappointing" appears rarely → ✅ more important

# 🧪 Code


re_list = [
    "good product",
    "good quality",
    "bad product"
]

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()

tfidf_matrix = tfidf.fit_transform(re_list)

print(tfidf_matrix.toarray())
print(tfidf.get_feature_names_out())


# 🎯 DONE ✅ (TF-IDF complete)
# 🚀 NEXT: Task 4 (Comparison Table)
# 📊 Comparison Table (Write this in assignment)
# Feature	OHE	BoW	TF-IDF
# Value	0 or 1	Count	Importance
# Meaning	Present or not	Frequency	Importance
# Problem	No frequency	No importance	No meaning understanding
# Use	Simple models	Basic NLP	Better models
# 🧠 Important Theory (Write this!)
# ❓ Which words are important in TF-IDF?

# 👉 Words that:

# appear less frequently
# but are unique

# Example:

# "disappointing" → high importance
# "good" → low importance
# ❓ Why common words get low weight?

# 👉 Because:

# They appear in many documents
# So they don’t help differentiate

# 👉 Example:

# "the", "good", "product" → everywhere
# → low value
# 🎯 Your Next Step

# Now do:

# 👉 1. Run BoW code
# 👉 2. Run TF-IDF code
# 👉 3. Print outputs

# Then tell me:

# 👉 Did you get matrices?

[[0.         0.70710678 0.70710678 0.        ]
 [0.         0.60534851 0.         0.79596054]
 [0.79596054 0.         0.60534851 0.        ]]
['bad' 'good' 'product' 'quality']


In [173]:
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = ["sky is blue", "sky is sky"]
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(corpus)

# Internally it is: <2x5 sparse matrix of type '<class 'numpy.float64'>'
print(type(tfidf_matrix)) 

# Now it is a dense numpy array
print(tfidf_matrix.toarray())


<class 'scipy.sparse._csr.csr_matrix'>
[[0.70490949 0.50154891 0.50154891]
 [0.         0.4472136  0.89442719]]


In [175]:
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = ["sky is blue", "sky is sky"]
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(corpus)
print(type(tfidf_matrix))
print(tfidf_matrix.toarray())

<class 'scipy.sparse._csr.csr_matrix'>
[[0.70490949 0.50154891 0.50154891]
 [0.         0.4472136  0.89442719]]


“A dense matrix ideally contains mostly non-zero values, while a sparse matrix contains mostly zeros. However, in practice, sparse data can still be stored in a dense format, which leads to memory inefficiency.”

🔍 Clarification
✅ Dense Matrix (Definition)

👉 Most values are non-zero

Example:

[1, 2, 3, 4]
[5, 6, 7, 8]
✅ Sparse Matrix (Definition)

👉 Most values are zero

Example:

[0, 0, 0, 1]
[0, 0, 2, 0]
⚠️ Important Confusion

👉 This:

[0, 0, 0, 1, 0, 0, 2, 0]

👉 is actually:
✅ Sparse data
❌ NOT dense

💡 So what is “dense matrix” then?

👉 Dense matrix = normal storage format

Even if data has many zeros, if we store like:

[0, 0, 0, 1, 0, 0, 2, 0]

👉 It is called:
➡️ Dense representation

🔥 Final Clear Difference
Type	Meaning
Sparse matrix (data)	mostly zeros
Dense matrix (data)	mostly non-zeros
Dense representation	stores ALL values
Sparse representation	stores ONLY non-zero values
🧠 In your case (NLP)

👉 Data:

Mostly zeros → Sparse data

👉 sklearn:

Uses sparse representation to save memory
🎯 Simple Memory Trick

👉 Sparse = many zeros (data)
👉 Dense = store everything (format)

💡 Correct Understanding (Final)

👉 Your statement:

“Dense matrix should not have many zeros”

✅ Correct (for data)

BUT

👉 In coding:

.toarray() → gives dense format
even if data has many zeros
🎯 Final Answer for Assignment

You can write:

“A dense matrix ideally contains mostly non-zero values, while a sparse matrix contains mostly zeros. However, in practice, sparse data can still be stored in a dense format, which leads to memory inefficiency.”

🚀 You’re doing excellent 👍

Now you have clear conceptual understanding — this is what interviewers want 💯

👉 Next:
Shall we finish

“Isn’t it opposite to their meaning?”

🧠 Short Answer

👉 No, it is NOT opposite
👉 It only feels confusing because of names

🔍 Why it feels opposite?

You are thinking:

👉 “Dense = should store only important values”
👉 “Sparse = should store many zeros”

👉 But actually, these names are about data structure, not logic.

✅ Correct Meaning
🔹 Dense Representation

👉 “Dense” means:
➡️ Everything is filled / stored

So:

[0, 0, 0, 1, 0, 0, 2, 0]

👉 We store:

all zeros
all non-zeros

✔️ Nothing is skipped
✔️ Fully filled → Dense

🔹 Sparse Representation

👉 “Sparse” means:
➡️ Data is mostly empty (zeros)

So we store only:

index: value
3: 1
6: 2

✔️ Skip zeros
✔️ Store only useful values

💡 Real-Life Analogy

Imagine a notebook 📒

Dense representation:

👉 You write every line, even if blank

Line1: 
Line2:
Line3: value
Line4:
Sparse representation:

👉 You write only useful lines

Line3: value
🎯 Key Understanding
Term	Meaning
Dense	store everything
Sparse	store only non-zero
⚠️ Important Clarification

👉 “Sparse” refers to:
➡️ data nature (many zeros)

👉 “Sparse representation” refers to:
➡️ smart storage method

🧠 Final Simple Line

👉 Dense = store all values
👉 Sparse = store only important (non-zero) values

🎯 Final Answer (for assignment)

You can write:

“Dense representation stores all values including zeros, whereas sparse representation stores only non-zero values to save memory. The terms may seem confusing, but they refer to storage methods rather than importance of values.”

🚀 You are doing excellent 👍

This level of doubt = strong understand

Comparison of OHE, BoW, and TF-IDF
Feature	OHE (One Hot Encoding)	BoW (Bag of Words)	TF-IDF
Meaning	Binary representation	Word frequency count	Importance of word
Values	0 or 1	Count (0,1,2..)	Weighted score
Order matters?	❌ No	❌ No	❌ No
Frequency used?	❌ No	✅ Yes	✅ Yes
Importance captured?	❌ No	❌ No	✅ Yes
Rare words importance-	❌ No	❌ No	✅ High
Common words- impact	Same as others	High	Low


or 






Task 4: Comparison Analysis

✅ 1. Comparison Table

You can directly write this in your assignment 👇

Feature	One Hot Encoding (OHE)	Bag of Words (BoW)	TF-IDF
Meaning	Only presence (0 or 1)	Counts word frequency	Importance of words
Value type	0 or 1	Integer (count)	Decimal (weight)
Example	good = 1	good = 2	good = 0.45
Repeated words	Not counted	Counted	Weighted
Common words	Same weight	Same weight	Low weight
Rare words	Same weight	Same weight	High weight
Use case	Simple models	Basic NLP	Advanced NLP



🧠 2. Which words are most important in TF-IDF?

👉 Rare words are most important

🔍 Example

Sentences:

good product
good quality
bad product

👉 Word frequency:

Word	Appears in sentences
good	2 times
product	2 times
bad	1 time
quality	1 time

👉 So:

bad, quality → HIGH importance ✅
good, product → LOW importance ❌
🧠 3. Why common words get LOW weight?

👉 Because they appear everywhere

🔍 Example

Word: “good”

Appears in many reviews
👉 Not special
👉 So TF-IDF reduces its value
🔍 Example

Word: “bad”

Appears rarely
👉 More meaningful
👉 So TF-IDF increases its value
🎯 Simple Explanation (Write this)

“In TF-IDF, rare words get higher importance because they carry more meaningful information, while common words get lower weight because they appear in many documents and are less informative.”

💡 One-line memory trick

👉 Common word = less important
👉 Rare word = more important

🧠 Final Understanding
Word type	TF-IDF weight
Common words	Low
Rare words	High
🚀 You finished Task 4 ✅
👉 Next Step

We go to Task 5 (Sparse Matrix Analysis):

Shape of matrix
Sparsity % calculation
Why inefficient